File: Common/InMotionGIT.Common/DataAcceess/Interface/ICommandExecution.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;

namespace InMotionGIT.Common.DataAccess.Interfaces
{

    public interface ICommandExecution
    {

        DataTable AsDataTable();

        Hashtable AsHashTable();

        List<T> AsList<T>();

        List<T> AsList<T>(ModelMapper<T> mapper);

        void ExecuteNonQuery();

        void ExecuteReader(Action<IDataReader> action);

        T ExecuteReaderSingle<T>(ModelMapper<T> mapper) where T : class, new();

        void ExecuteReaderSingle(Action<IDataReader> action);

        T ExecuteScalar<T>();

        T ExecuteScalar<T>(T defaultValue);

        ICommandExecution WithParam(string paramName, object paramValue);

    }

}

File: Common/InMotionGIT.Common - Copy/DataAcceess/Interface/ICommandExecution.vb
Namespace DataAccess.Interfaces

    Public Interface ICommandExecution

        Function AsDataTable() As DataTable

        Function AsHashTable() As Hashtable

        Function AsList(Of T)() As List(Of T)

        Function AsList(Of T)(ByVal mapper As ModelMapper(Of T)) As List(Of T)

        Sub ExecuteNonQuery()

        Sub ExecuteReader(ByVal action As Action(Of IDataReader))

        Function ExecuteReaderSingle(Of T As {Class, New})(ByVal mapper As ModelMapper(Of T)) As T

        Sub ExecuteReaderSingle(ByVal action As Action(Of IDataReader))

        Function ExecuteScalar(Of T)() As T

        Function ExecuteScalar(Of T)(ByVal defaultValue As T) As T

        Function WithParam(ByVal paramName As String, ByVal paramValue As Object) As ICommandExecution

    End Interface

End Namespace